### PR TITLE
Feature/#303 Design Timeout DAO layer and flow

### PR DESCRIPTION
**Pull request by ggrg**
Documented the output of story #303 as design/ markdown document. This design is to be later implemented in central-ledger with Node.js/knex. This PR is connected to the same name PR against central-ledger [here](https://github.com/mojaloop/central-ledger/pull/105).

### CODE REVIEW

File: CentralServices/designs/des-timeout-1.0.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+```
+As a result of calling `RandomTransfer.insert(config)` we got 1M transfers in the *transfer* table and approximatelly 3.5M records in *transferStateChange* table. The number of expired transfers in *RECEIVED_PREPARE* or *RESERVED* state was 1000 at the time of the generation of the transfers, but this number was quickly growing as time was passing and other randomly inserted transfers' expirationDate got passed by the present time. A quick fix to that is to shift the expirationDate of all 1M transfers in table *transfer* by the interval of time that passed since data generation:
+```sql
+UPDATE central_ledger.transfer t
+SET expirationDate = expirationDate + interval 1 hour;
+```
+
+### DB Model Changes
+
+We currently don't update the transferState, but make insert for each state change instead. As a result we have multiple records in transferStateChange table which correspond to every transfer.
```
Comment (rmothilal) [inline CentralServices/designs/des-timeout-1.0.md]: is this supposed to be transferStateChange?

File: CentralServices/designs/des-timeout-1.0.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+SET expirationDate = expirationDate + interval 1 hour;
+```
+
+### DB Model Changes
+
+We currently don't update the transferState, but make insert for each state change instead. As a result we have multiple records in transferStateChange table which correspond to every transfer. 
+
+In order to get the current state of any transfer we need to select the last inserted record. When working with single transfer this is achieved with descending ordering by id and getting the first record. But when we want to do so for a million transfers we need to use *INNER JOIN* of subquery with *MAX* function applied over *GROUP BY transferId*. 
+
+After running such query on transferStateChange table for 3.5M it resulted in nearly 60s execution time (before joining the result to the other part of the query). It was clear it is not satisfactory and we couldn't scale for the future data growth. The solution we found was to proceed only the newly inserted transfer state changes. In this way the *GROUP BY* subquiry won't be run over such large datasets of 3.5M, but just over the dataset of newly inserts after the previous run of the TimeoutHandler.
```
Comment (rmothilal) [inline CentralServices/designs/des-timeout-1.0.md]: say 60 seconds not 60s

Comment (rmothilal) [reply CentralServices/designs/des-timeout-1.0.md]: subquery not subquiry

File: CentralServices/designs/des-timeout-1.0.md (hunk 1 above)
Comment (ggrg) [reply CentralServices/designs/des-timeout-1.0.md]: You're correct but it's actually meant not to be table name, so I'll change it to transfer state. Thanks.

File: CentralServices/designs/des-timeout-1.0.md (hunk 2 above)
Comment (ggrg) [reply CentralServices/designs/des-timeout-1.0.md]: Changed, thanks. Also I've run a spell checker on the text using Pages and found some other typos.

**Pull request merged**